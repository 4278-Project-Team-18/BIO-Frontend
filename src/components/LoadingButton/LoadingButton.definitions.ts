/* eslint-disable autofix/no-unused-vars */
import type { IconDefinition } from "@fortawesome/fontawesome-svg-core";

export interface LoadingButtonProps {
  isLoading: boolean;
  text: string;
  onClick?: () => void;
  isLoadingText?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  styles?: any;
  icon?: IconDefinition;
  variant?: LoadingButtonVariant;
  type?: "button" | "submit" | "reset";
  disabled?: boolean;
  isResponsive?: boolean;
  iconOnly?: boolean;
}

export enum LoadingButtonVariant {
  GREEN = "green",
  RED = "red",
  BLUE = "blue",
  YELLOW = "yellow",
  GREY = "grey",
  PURPLE = "purple",
  DARKGREY = "darkgrey",
  BLACK = "black",
}

export const LoadingButtonVariantColorMap = {
  [LoadingButtonVariant.GREEN]: "#5ad06d",
  [LoadingButtonVariant.RED]: "#EF4444",
  [LoadingButtonVariant.BLUE]: "#3B82F6",
  [LoadingButtonVariant.YELLOW]: "#F59E0B",
  [LoadingButtonVariant.GREY]: "#CCCCCC",
  [LoadingButtonVariant.PURPLE]: "#8B5CF6",
  [LoadingButtonVariant.DARKGREY]: "#AAAAAA",
  [LoadingButtonVariant.BLACK]: "#222222",
};
