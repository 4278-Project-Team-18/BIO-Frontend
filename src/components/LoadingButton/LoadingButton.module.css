.loading-button {
  width: min-content;
  height: 40px;
  border-radius: 10px;
  background-color: #54d469;
  color: white;
  font-size: 16px;
  font-weight: 600;
  border: none;
  cursor: pointer;
  box-shadow: 0px 3px 5px rgba(0, 0, 0, 0.2);
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: row;
}

.loading-button-spinner {
  font-size: 16px;
  margin-left: 5px;
  margin-right: 5px;
  color: #ffffff;
}

@media (max-width: 700px) {
  .loading-button-label-responsive {
    display: none;
  }

  .loading-button-label {
    font-weight: 600;
    font-size: 16px;
    color: #ffffff;
    font-family: "Poppins", sans-serif;
    font-weight: 600;
    width: fit-content;
    padding-left: 5px;
    padding-right: 5px;
    white-space: nowrap;
  }
}

@media (min-width: 701px) {
  .loading-button-label {
    font-weight: 600;
    font-size: 16px;
    color: #ffffff;
    font-family: "Poppins", sans-serif;
    font-weight: 600;
    width: fit-content;
    padding-left: 5px;
    padding-right: 5px;
    white-space: nowrap;
  }

  .loading-button-label-responsive {
    font-weight: 600;
    font-size: 16px;
    color: #ffffff;
    font-family: "Poppins", sans-serif;
    font-weight: 600;
    width: fit-content;
    padding-left: 5px;
    padding-right: 5px;
    white-space: nowrap;
  }
}
