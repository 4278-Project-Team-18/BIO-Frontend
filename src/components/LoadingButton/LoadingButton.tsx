import styles from "./LoadingButton.module.css";
import {
  LoadingButtonVariant,
  type LoadingButtonProps,
  LoadingButtonVariantColorMap,
} from "./LoadingButton.definitions";
import { MoonLoader } from "react-spinners";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

/**
 * A button that shows a loading spinner when loading
 *
 * @param text - REQUIRED: the text to display on the button
 * @param isLoading - REQUIRED: whether or not the button is loading
 * @param onClick - optional function to call when the button is clicked
 * @param styles - optional styles to apply to the button
 * @param isLoadingText - optional text to display when the button is loading (defaults to "Loading...")
 * @param icon - optional icon to display on the button next to label text (defaults to nothing)
 * @param variant - optional variant of the button (defaults to green)
 * @param type - optional type of the button (defaults to "button")
 */
const LoadingButton = ({
  onClick,
  text,
  isLoading,
  isLoadingText = "Loading...",
  styles: style,
  icon,
  variant = LoadingButtonVariant.GREEN,
  type = "button",
}: LoadingButtonProps) => (
  <button
    onClick={onClick}
    className={styles["loading-button"]}
    style={{
      ...style,
      backgroundColor: LoadingButtonVariantColorMap[variant],
    }}
    type={type}
  >
    <span className={styles["loading-button-label"]}>
      {isLoading ? isLoadingText : text}
    </span>
    {icon && !isLoading ? (
      <FontAwesomeIcon
        icon={icon}
        className={styles["loading-button-spinner"]}
      />
    ) : (
      <MoonLoader
        size={16}
        color="#fff"
        loading={isLoading}
        className={styles["loading-button-spinner"]}
      />
    )}
  </button>
);

export default LoadingButton;
