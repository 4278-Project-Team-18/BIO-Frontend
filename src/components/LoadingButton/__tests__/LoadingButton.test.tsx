import LoadingButton from "../LoadingButton";
import { render } from "@testing-library/react";

describe("LoadingButton Tests", () => {
  test("LoadingButton should render correctly", () => {
    const mockProps = {
      isLoading: true,
      isLoadingText: "Loading Test",
      text: "Test",
      onClick: jest.fn(),
    };

    const component = render(<LoadingButton {...mockProps} />);
    expect(component).toBeTruthy();

    // Check if the loading text is rendered
    expect(component.getByText(mockProps.isLoadingText)).toBeTruthy();
    component.getByText(mockProps.isLoadingText).click();
    expect(mockProps.onClick).toHaveBeenCalled();

    // Check if text is rendered
    mockProps.isLoading = false;
    const component2 = render(<LoadingButton {...mockProps} />);
    expect(component2.getByText(mockProps.text)).toBeTruthy();
  });

  test("LoadingButton snapshot should render correctly", () => {
    const mockProps = {
      isLoading: true,
      isLoadingText: "Loading Test",
      text: "Test",
      onClick: jest.fn(),
    };

    const component = render(<LoadingButton {...mockProps} />);
    expect(component).toMatchSnapshot();
  });
});
