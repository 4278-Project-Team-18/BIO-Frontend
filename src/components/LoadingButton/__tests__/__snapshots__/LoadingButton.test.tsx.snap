// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingButton Tests LoadingButton snapshot should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="loading-button"
        style="background-color: rgb(90, 208, 109);"
        type="button"
      >
        <span
          class="loading-button-label"
        >
          Loading Test
        </span>
        <span
          class="loading-button-spinner"
          style="display: inherit; position: relative; width: 20.57142857142857px; height: 20.57142857142857px; animation: react-spinners-MoonLoader-moon 0.6s 0s infinite linear; animation-fill-mode: forwards;"
        >
          <span
            style="width: 2.2857142857142856px; height: 2.2857142857142856px; border-radius: 100%; background-color: rgb(255, 255, 255); opacity: 0.8; position: absolute; top: 6.857142857142858px; animation: react-spinners-MoonLoader-moon 0.6s 0s infinite linear; animation-fill-mode: forwards;"
          />
          <span
            style="width: 16px; height: 16px; border-radius: 100%; border: 2.2857142857142856px solid #fff; opacity: 0.1; box-sizing: content-box; position: absolute;"
          />
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="loading-button"
      style="background-color: rgb(90, 208, 109);"
      type="button"
    >
      <span
        class="loading-button-label"
      >
        Loading Test
      </span>
      <span
        class="loading-button-spinner"
        style="display: inherit; position: relative; width: 20.57142857142857px; height: 20.57142857142857px; animation: react-spinners-MoonLoader-moon 0.6s 0s infinite linear; animation-fill-mode: forwards;"
      >
        <span
          style="width: 2.2857142857142856px; height: 2.2857142857142856px; border-radius: 100%; background-color: rgb(255, 255, 255); opacity: 0.8; position: absolute; top: 6.857142857142858px; animation: react-spinners-MoonLoader-moon 0.6s 0s infinite linear; animation-fill-mode: forwards;"
        />
        <span
          style="width: 16px; height: 16px; border-radius: 100%; border: 2.2857142857142856px solid #fff; opacity: 0.1; box-sizing: content-box; position: absolute;"
        />
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
